libsplinter/src/network/peer_manager/interconnect.rs

```diff
@@ -0,0 +1,650 @@
+        self
+    }
+
+    /// Add a MatrixShutdown to PeerInterconnectBuilder
+    ///
+    /// # Arguments
+    ///
+    /// * `message_shutdown` - a MatrixShutdown that will be used to shutdown a MatrixSender
+    pub fn with_message_shutdown(mut self, message_shutdown: V) -> Self {
+        self.message_shutdown = Some(message_shutdown);
+        self
+    }
+
+    /// Add a Dispatcher for NetworkMessageType to PeerInterconnectBuilder
+    ///
+    /// # Arguments
+    ///
+    /// * `network_dispatcher` - a Dispatcher that is loaded with the Handlers for
+    ////       NetworkMessageTypes
+    pub fn with_network_dispatcher(
+        mut self,
+        network_dispatcher: Dispatcher<NetworkMessageType>,
+    ) -> Self {
+        self.network_dispatcher = Some(network_dispatcher);
+        self
+    }
+
+    /// Build the PeerInterconnect. This function will build the dispatch loop for network message
+    /// types and start up threads to send and recv messages from the peers.
+    ///
+    /// Returns the PeerInterconnect object that can be used to get network message senders and
+    /// shutdown message threads.
+    pub fn build(&mut self) -> Result<PeerInterconnect<V>, PeerInterconnectError> {
+        let (dispatched_sender, dispatched_receiver) = channel();
+        let peers = Arc::new(Mutex::new(BiHashMap::new()));
+
+        let peer_connector = self.peer_connector.take().ok_or_else(|| {
+            PeerInterconnectError::StartUpError("Peer manager connecor missing".to_string())
+        })?;
+
+        let message_shutdown = self.message_shutdown.take().ok_or_else(|| {
+            PeerInterconnectError::StartUpError("Message shutdown missing".to_string())
+        })?;
+
+        let mut network_dispatcher = self.network_dispatcher.take().ok_or_else(|| {
+            PeerInterconnectError::StartUpError("Network dispatcher missing".to_string())
+        })?;
+        network_dispatcher.set_network_sender(NetworkMessageSender::new(dispatched_sender.clone()));
+        let network_dispatch_loop = DispatchLoopBuilder::new()
+            .with_dispatcher(network_dispatcher)
+            .build()
+            .map_err(|err| {
+                PeerInterconnectError::StartUpError(format!(
+                    "Unable to start network dipatch loop: {}",
+                    err
+                ))
+            })?;
+        let network_dispatcher_sender = network_dispatch_loop.new_dispatcher_sender();
+        let network_dispatcher_shutdown = network_dispatch_loop.shutdown_signaler();
+
+        // start receiver loop
+        let message_receiver = self.message_receiver.take().ok_or_else(|| {
+            PeerInterconnectError::StartUpError("Message receiver missing".to_string())
+        })?;
+
+        let recv_peers = peers.clone();
+        let recv_peer_connector = peer_connector.clone();
+        let recv_join_handle = thread::Builder::new()
+            .name("Peer Interconnect Receiver".into())
```

Comment:
Do we want to enforce consistency on thread naming? I think some other examples are camel case?

Reply:
Based on a quick search most threads in libsplinter are named with spaces in the name

Examples:
 - Connection Manager
 - Peer Manager
 - Orchestrator Incoming
 - Orchestrator Outgoing 

But it is a mixed bag. I will add a story to the backlog for standardizing thread names.